linode/manager
# upcoming: [DI-20348] - Changes for Adding CloudPulseCustomSelect component and Integration

## Description 📝
Changes for new component CloudPulseCustomSelect that helps to display various filters in the dashboard view and integration of custom select component with metrics API call.

## Changes  🔄
List any change relevant to the reviewer.
1. Created new component CloudPulseCustomSelect.
2. Integration of the select custom filters with the metrics API call.

## Target release date 🗓️
9/3/24

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
|![Screenshot 2024-08-21 at 12 55 32 PM](https://github.com/user-attachments/assets/e9647a69-fa8f-49a2-b248-b635faaf93e5)|![Screenshot 2024-08-21 at 12 55 51 PM](https://github.com/user-attachments/assets/259fb5c9-1cf8-4eaa-94eb-64bb85cbdd5e)|
|![Screenshot 2024-08-21 at 12 55 59 PM](https://github.com/user-attachments/assets/771fb77c-e660-477f-8c59-96ea697f013d)|![Screenshot 2024-08-21 at 12 56 12 PM](https://github.com/user-attachments/assets/12a20d2a-b728-4b06-828f-4f8bc0a48f99)|
|![Screenshot 2024-08-21 at 12 56 23 PM](https://github.com/user-attachments/assets/a85107b2-2aee-4653-b719-1e98cd169393)|![Screenshot 2024-08-21 at 5 19 25 PM](https://github.com/user-attachments/assets/6e356a22-fdc0-41d4-81a0-32740e5f9078)|
![Screenshot 2024-08-21 at 1 02 02 PM](https://github.com/user-attachments/assets/acc8b95a-87e9-4328-a153-8074594813eb)|
|![Screenshot 2024-08-21 at 1 10 03 PM](https://github.com/user-attachments/assets/338712fb-1a79-48cb-b951-aaf9f52d737e)|![Screenshot 2024-08-21 at 1 10 35 PM](https://github.com/user-attachments/assets/df58c334-90cb-493f-a225-48544b0e9cf2)|


## How to test 🧪
1. Login as mock user since some of the API endpoints are not available.
2. Go the Monitors tab.
3. In the dashboard selection dropdown, select Dbaas Dashboard.
4. In the Filters selection section, you can see select an Engine dropdown, that uses CloudPulseCustomSelect component.
5. Also you can add a filter in the FilterConfig.ts file for dbaas like below and upon selection, we can see the filter goes in the metrics API request inside filters object.
```
{
      configuration: {
        filterKey: 'test',
        filterType: 'string',
        isFilterable: true, // isFilterable -- this determines whethere you need to pass it metrics api
        isMetricsFilter: false, // if it is false, it will go as a part of filter params, else global filter
        isMultiSelect: false,
        name: 'Test Filter',
        neededInServicePage: false,
        options: [
          {
            id: 'test',
            label: 'Test1',
          },
          {
            id: 'test2',
            label: 'Test2',
          },
        ],
        placeholder: 'Select a Test',
        priority: 5,
        type: CloudPulseSelectTypes.static,
      },
      name: 'Test Filter',
    }
```
6. In the above config, if you remove the options, the error message will appear below the dropdown
**Note-** You might encounter flickering issues, that is caused due to preferences, that will be solved in the upcoming PR

## As an Author I have considered 🤔

*Check all that apply*
